### PR TITLE
MetaT update to JGI workflows

### PR DESCRIPTION
**Pull request by kaijli**
Workflows all imported via https except for json output creation. Updated supporting files and cleaned up files no longer used. Successful expanse run at `ed1aa6ed-b18c-4e12-85e2-47d1ce569531`, working on successful jaws run. Otherwise ready for importing to nmdc-edge.

Some things that could be changed if needed:
 - interleaved vs paired file handling 
 - output file selection, renaming?

### CODE REVIEW

**Comment by aclum** [comment]
@kaijli  can you fix the merge conflicts?

File: README.md [hunk 1]
```diff
@@ -47,64 +51,38 @@ java -Dconfig.file=wdls/shifter.conf -jar /full/path/to/cromwell-XX.jar run -i i
...
 
-- `microbiomedata/meta_t:latest`. 
-`Dockerfile` can be found in `Docker/metatranscriptomics/` directory. 
-- `microbiomedata/bbtools:38.94`
-- `scanon/nmdc-meta:v0.0.1`
-- `bfoster1/img-omics:0.1.7`
-- `scanon/im-trnascan:v0.0.1`
-- `scanon/im-last:v0.0.1`
-- `scanon/im-hmmsearch:v0.0.0`
+- `microbiomedata/meta_t:latest`
```
Comment (aclum) [inline README.md:54]: We should use a tagged image here, not latest

File: to_json.wdl [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+		File cds_json = "~{prefix}_cds_counts.json"
+		File sense_json = "~{prefix}_sense_counts.json"
+		File anti_json = "~{prefix}_antisense_counts.json"
+        File top100_json = "~{prefix}_top100_features.json"
+		File sorted_json = "~{prefix}_sorted_features.json"
+        File sorted_tsv = "~{prefix}_sorted_features.tsv"
+	}
+
+	runtime {
+		docker: container
```
Comment (aclum) [inline to_json.wdl:233]: add memory and cpu to the runtime block

File: wdls/.metat_tasks.wdl.swp
Comment (aclum) [inline wdls/.metat_tasks.wdl.swp:1]: can this file be deleted?

File: to_json.wdl [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+												"stdevA"])
+			antisense_reads = cds_only[cds_only['strand'] == "-"].drop(columns = ["sense_read_count", 
+												"mean",
+												"median",
+												"stdev"])
+
+			sorted_features = gff_rc_pd.sort_values(by='sense_read_count', ascending=False)
+			top100 = sorted_features[:100]
+			
+			write_json(gff_rc_pd.to_dict(orient="records"), gff_rc_json)
```
Comment (mflynn-lanl) [inline to_json.wdl:83]: You could use https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_json.html

File: to_json.wdl [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+												keep_order=True,
+												merge_strategy="create_unique")
+					print("New gffutils db created")
+				else:
+					gtf_as_db = gffutils.FeatureDB("metat_db.db", keep_order=True)
+					print("Cached gffutils db loaded")
+				json_list = []
+				for feat_obj in gtf_as_db.all_features():
+					feat_dic = {}  # an empty dictionary to append features
+					feat_dic_str = self.collect_features(feat_obj=feat_obj, feat_dic=feat_dic)
```
Comment (mflynn-lanl) [inline to_json.wdl:135]: I think it is not recommended to pass an empty dict as input to a function: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

File: README.md (hunk 1 above)
Comment (kaijli) [reply README.md:54]: I didn't see any tags on the docker repo, we'll have to make a new one https://hub.docker.com/r/microbiomedata/meta_t/tags

File: to_json.wdl (hunk 2 above)
Comment (aclum) [reply to_json.wdl:233]: @chienchi  can you take care of this? This was your image from 2 years ago.

File: wdls/.metat_tasks.wdl.swp
Comment (kaijli) [reply wdls/.metat_tasks.wdl.swp:1]: the .swp file or the tasks? I'll bring the tasks back and update to v1.0

Comment (kaijli) [reply wdls/.metat_tasks.wdl.swp:1]: brought it back and added to_json task
https://github.com/microbiomedata/metaT/blob/workflows/metat_tasks.wdl

File: README.md (hunk 1 above)
Comment (kaijli) [reply README.md:47]: @chienchi  i think alicia meant to mention you here. I've tried to build the docker image from the commands at the top of the dockerfile, but I run into errors with `continuumio/miniconda3:4.8.2`

```
ERROR: failed to solve: process "/bin/sh -c apt-get update -y     && apt-get install -y build-essential unzip wget curl gawk locales    && apt-get clean" did not complete successfully: exit code: 100
```

I was able to build an image by bumping the version to `continuumio/miniconda3:4.10.3` but I don't have push permissions to the docker repo
https://github.com/microbiomedata/metaT/blob/workflows/docker/Dockerfile
@aclum it seems like you have push permissions as well, if you have the time to make/push the image

Comment (chienchi) [reply README.md:47]: I tried to build the image but it seems `conda` takes long time.  @kaijli I added the meta_t docker repo into developer team on docker hub which you are member of. Please try to push the updated image. However, question here is "do we still need this image after update to JGI workflow?"

File: to_json.wdl (hunk 2 above)
Comment (chienchi) [reply to_json.wdl]: added https://github.com/microbiomedata/metaT/pull/36/commits/08cb4e1fc49c20fd041a1f6b5d61919cfb521b85

File: README.md (hunk 1 above)
Comment (kaijli) [reply README.md:47]: @chienchi I was trying to build a new image with just these python packages for the json tasks, but kept running into issues with my dockerfile

```
channels:
  - conda-forge
  - bioconda
dependencies:
  - pandas=2.1.4
  - gffutils=0.13
  - python=3.11.0
  ```
  python version can be lower, it runs on 3.7.12

**Comment by kaijli** [comment]
successful jaws run at `81671` and expanse run at `6124aa02-1ddc-4879-a078-fe2336139e2e` with requested edits

File: metaT.wdl [hunk 5]
```diff
@@ -1,21 +1,23 @@
 # metaT workflow wrapper
 version 1.0
 
-import "https://raw.githubusercontent.com/microbiomedata/metaT_ReadsQC/v0.0.1/rqcfilter.wdl?ref=b9f57348975e0ac4a32f733153ff579488e55ba9" as readsqc
-import "https://raw.githubusercontent.com/microbiomedata/metaT_Assembly/v0.0.0/metaT_assembly.wdl?ref=75f6272246afc4b38f080f844752723e2094634a" as assembly
+import "https://raw.githubusercontent.com/microbiomedata/metaT_ReadsQC/main/rqcfilter.wdl" as readsqc
```
Comment (aclum) [inline metaT.wdl:4]: we should be using tagged versions here.

File: metat_tasks.wdl [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+      cp ~{crt_gff} ~{annodir}/~{prefix}_crt.gff
+      cp ~{genemark_gff} ~{annodir}/~{prefix}_genemark.gff
+      cp ~{prodigal_gff} ~{annodir}/~{prefix}_prodigal.gff
+      cp ~{trna_gff} ~{annodir}/~{prefix}_trna.gff
+      cp ~{rfam_gff} ~{annodir}/~{prefix}_rfam.gff
+      cp ~{product_names_tsv} ~{annodir}/~{prefix}_product_names.tsv
+      cp ~{crt_crisprs} ~{annodir}/~{prefix}_crt.crisprs
+      cp ~{imgap_version} ~{annodir}/~{prefix}_anno.info
+      cp ~{renamed_fasta} ~{annodir}/~{prefix}_renamed_contigs.fna
+      cp ~{map_file} ~{annodir}/~{prefix}_map.fna
```
Comment (aclum) [inline metat_tasks.wdl:572]: This file is incorrectly named nmdc_xxxxxxx_map.fna, this should be _contig_names_mapping.tsv

File: metat_tasks.wdl [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+         File final_trna_gff = "~{annodir}/~{prefix}_trna.gff"
+         File final_rfam_gff = "~{annodir}/~{prefix}_rfam.gff"
+         File final_product_names_tsv = "~{annodir}/~{prefix}_product_names.tsv"
+         File final_crt_crisprs = "~{annodir}/~{prefix}_crt.crisprs"
+         File final_imgap_version = "~{annodir}/~{prefix}_anno.info"
+         File final_renamed_fasta = "~{annodir}/~{prefix}_renamed_contigs.fna"
+         File final_map_file = "~{annodir}/~{prefix}_map.fna"
+
+         # metaT_ReadCounts
+         File final_count_table = "~{readmap}/~{prefix}.readcount"
```
Comment (aclum) [inline metat_tasks.wdl:641]: suggest naming this with the same suffix as JGI does `.rnaseq_gea.txt`

File: metat_tasks.wdl [hunk 8]
```diff
@@ -0,0 +1,665 @@
...
+         File? final_count_ig = "~{readmap}/~{prefix}.readcount.intergenic"
```
Comment (aclum) [inline metat_tasks.wdl:642]: suggest naming this with the same suffix as JGI does `.rnaseq_gea.intergenic.txt`

File: README.md (hunk 1 above)
Comment (kaijli) [reply README.md]: update to close thread: was able to make tagged releases for this new workflow, will need to make new images in future, tagged images have issue with old workflow somehow

**Review by aclum — APPROVED** [review]

**Comment by aclum** [comment]
@kaijli  does this still need to be merged? Or was this handled in another merge and should be closed w/o merging?

**Pull request merged**